File: Chapter 4/Ruby/product_validation_test.rb
require 'rubygems'
require 'selenium-webdriver'
require 'test/unit'   
require File.join(File.dirname(__FILE__), 'test_data')

class ProductValidationTests < Test::Unit::TestCase  
  def setup  
    require 'debug'
    @selenium = Selenium::WebDriver.for(:firefox)         
  end

  def teardown
     @selenium.quit
  end

  def test_all_products_with_fixtures     
     TestData.get_product_fixtures.values.each do | product_info |
       @selenium.get(TestData.get_base_url + product_info["url"])        
       verify_product_info(product_info)     
     end    
  end
  
  def test_all_products_with_api_response      
    TestData.get_products_from_api.each do |product_info|
       @selenium.get(TestData.get_base_url + product_info["url"])          
       verify_product_info(product_info)
    end                              
  end            
 
  private
  
  def verify_product_info(product_info)
     failure_info = "Product Name: #{product_info['name']} \nPermalink #{product_info['url']}\n"
     assert_equal(product_info["name"], get_product_title, failure_info)  
     assert_equal("#{TestData.get_base_url + product_info["url"].gsub(/^\//, "")}/", get_current_url, failure_info)
     assert_equal(product_info["description"], get_product_description, failure_info)
     assert_equal(product_info["price"].to_s, get_product_price, failure_info)
  end
  
  def get_product_title
    @selenium.find_element(:class, "category-title").text 
  end 
  
  def get_product_description
    @selenium.find_element(:id, "main-products").find_element(:class, "content").text
  end 
  
  def get_product_price
    @selenium.find_element(:class, "price-tag").text.gsub(/[\n\$]/, "")
  end
  
  def get_current_url
    @selenium.current_url
  end
   
end

File: Java/src/test/java/chapter5/ProductValidationTest.java
package chapter5;

import com.google.gson.internal.StringMap;

import org.junit.Before;
import org.junit.Test;
import org.junit.After;

import org.openqa.selenium.By;

import java.util.Map;


import chapter5.fixtures.TestData;
import chapter5.helpers.SeleniumWrapper;

import static org.junit.Assert.assertEquals;


public class ProductValidationTest {

  private SeleniumWrapper selenium;

  @After
  public void tearDown() throws Exception {
    selenium.quit();
  }

  @Before
  public void setUp() throws Exception {
    selenium = new SeleniumWrapper("firefox");
  }

  @Test
  public void testAllProductsWithFixtures() throws Exception {
    Map<String, Object> foo = TestData.getProductFixtures();
    for ( String key : foo.keySet()) {
      Map<String, Object> productInfo = (Map<String, Object>) foo.get(key);
      selenium.get(TestData.getBaseUrl() + productInfo.get("url"));
      verifyProductInfo(productInfo);
    }
  }

  @Test
  public void testAllProductsWithApiResponse() throws Exception {
    for (StringMap productInfo : TestData.getProductsFromApi()) {
      selenium.get(TestData.getBaseUrl() + productInfo.get("url"));
      verifyProductInfo(productInfo);
    }
  }

  private void verifyProductInfo(Map productInfo) {
    assertEquals(productInfo.get("name"), getProductTitle());
    String
        expectedUrl = TestData.getBaseUrl() + ((String) productInfo.get("url")).replaceAll("/", "") + "/";
    assertEquals(expectedUrl, getCurrentUrl());
    assertEquals(productInfo.get("description"), getProductDescription());
    assertEquals(productInfo.get("price").toString(), getProductPrice());
  }

  private String getProductPrice() {
    return selenium.getInnerText(By.className("price-tag")).replaceAll("[\n\\$]", "");
  }

  private String getProductDescription() {
    return selenium.findElement(By.id("main-products")).findElement(By.className("content"))
        .getText();
  }

  private String getProductTitle() {
    return selenium.getInnerText(By.className("category-title"));
  }

  public String getCurrentUrl() {
    return selenium.getCurrentUrl();
  }
}